doc/DESIGN.md

```diff
@@ -0,0 +1,1625 @@
+optimizations. The translation layer translates RealRegs as particular
+vregs at the regalloc2 layer, because we need to track their liveness
+properly. Handling these as "normal" vregs, with massive bundles of
+many liveranges throughout the function, turns out to be a very
+inefficient solution. So we mark them as "pinned" with a hook in the
+RA2 API. Semantically, this means they are always assigned to a
+particular preg whenever mentioned in an operand (but *NOT* between
+those points; it is possible for a pinned vreg to move all about
+registers and stackslots as long as it eventually makes it back to its
+home preg in time for its next use).
+
+This has a few implications during liverange construction. First, when
+we see an operand that mentions a pinned vreg, we translate this to an
+operand constraint that names a fixed preg. Later, when we build
+bundles, we will not create a bundle for the pinned vreg; instead we
+will transfer its liveranges directly as unmoveable reservations in
+pregs' allocation maps. Finally, we need to handle moves specially.
+
+With the caveat that "this is a massive hack and I am very very
+sorry", here is how it works. A move between two pinned vregs is easy:
+we add that to the inserted-moves list right away because we know the
+Allocation on both sides. A move from a pinned vreg to a normal vreg
+is the first interesting case. In this case, we (i) create a ghost def
+with a fixed-register policy on the normal vreg, doing the other
+liverange-maintenance bits as above, and (ii) adjust the liveranges on
+the pinned vreg (so the preg) in a particular way. If the preg is live
+flowing downward, then this move implies a copy, because the normal
+vreg and the pinned vreg are both used in the future and cannot
+overlap. But we cannot keep the preg continuously live, because at
+exactly one program point, the normal vreg is pinned to it. So we cut
+the downward-flowing liverange just *after* the normal vreg's
+fixed-reg ghost def. Then, whether it is live downward or not, we
+create an upward-flowing liverange on the pinned vreg that ends just
+*before* the ghost def.
+
+The move-from-normal-to-pinned case is similar. First, we create a
+ghost use on the normal vreg that pins its value at this program point
+to the fixed preg. Then, if the preg is live flowing downward, we trim
+its downward liverange to start just after the fixed use.
+
+There are also some tricky metadata-maintenance records that we emit
+so that the checker can keep this all straight.
+
+The outcome of this hack, together with the operand-constraint
+translation on normal uses/defs/mods on pinned vregs, is that we
+essentially are translating regalloc.rs's means of referring to real
+registers to regalloc2's preferred abstractions by doing a bit of
+reverse-engineering. It is not perfect, but it works. Still, we hope
+to rip it all out once we get rid of the need for the compatibility
+shim.
```

Comment:
Ah, yes, there should always be a way to "pass through" a non-allocatable register.

Actually I hadn't thought of exactly how this could work more efficiently, as the native-regalloc2-API answer would have been to use a fresh vreg and constrain it with an `OperandPolicy::Fixed(preg)`, but that actually doesn't work for uses (it would flow upward to entry and be an unresolved live-in).

I think that for a non-allocatable register, i.e. one that does not need to track liveranges or anything else from the regalloc POV, we could just declare that an `Operand` with `VReg::invalid()` and `OperandPolicy::Fixed(preg)` causes the regalloc to immediately assign `Allocation::reg(preg)` to the corresponding slot and then ignore it. I can go ahead and implement that now.

Thanks for raising this!

Reply:
Added in latest commit.